File: Data Structures & Algorithms/binary-tree-right-side-view/submission-13.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        std::queue<TreeNode*> q;
        std::vector<int> res;
        q.push(root);

        while (!q.empty()) {
            int qLen = q.size();
            int right = INT_MAX;

            for (int i = 0; i < qLen; i++) {
                TreeNode* cur = q.front();
                q.pop();

                if (cur) {
                    right = cur->val;
                    q.push(cur->left);
                    q.push(cur->right);
                }
            }

            if (right != INT_MAX) {
                res.push_back(right);
            }
        }

        return res;
    }
};

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-13.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        def isSameTree(main, sub):
            if not main and not sub:
                return True

            if main and sub and main.val == sub.val:
                return isSameTree(main.left, sub.left) and isSameTree(main.right, sub.right)

            else:
                return False

        if not root:
            return False

        if not subRoot:
            return True

        if isSameTree(root, subRoot):
            return True
            
        return self.isSubtree(root.right, subRoot) or self.isSubtree(root.left, subRoot)